### PR TITLE
Enable T0 and sample position calibration in SCDCalibratePanels

### PR DESCRIPTION
**Pull request by KedoKudo**
**Description of work.**

- Original Gitlab User Story: [SCD316](https://code.ornl.gov/sns-hfir-scse/diffraction/single-crystal/single-crystal-diffraction/-/issues/316)
- This is #31514 into `master`

This PR introduces the following improvements for `SCDCalibratePanels`:

- Users can calibrate `T0` and tune `sample-position` if these two are included in the calibration routine.
- Each calibration field has its own search radius and tolerance (default values provided are based on TOPAZ).
- A new output parameter, `T0` is added so that users can directly use it from memory instead of parsing the calibration output file (detcal and xml both contain T0 information)
- The output parameter file (XML file) now contains two new section: T0 and sample position.

**To test:**

- Using the following script to generate a synthetic peaksworkspace, perturb the instrument geometry, and the `T0` shift

```python
# import mantid algorithms, numpy and matplotlib
from mantid.simpleapi import *
from mantid.geometry import CrystalStructure
import matplotlib.pyplot as plt
import numpy as np

from collections import namedtuple

def convert(dictionary):
    return namedtuple('GenericDict', dictionary.keys())(**dictionary)

# --- setup simulated workspace
# data from Mantid web documentation
lc_natrolite = {
    "a": 18.29,  # A
    "b": 18.64,  # A
    "c": 6.56,  # A
    "alpha": 90,  # deg
    "beta": 90,  # deg
    "gamma": 90,  # deg
}
natrolite = convert(lc_natrolite)

# Generate simulated workspace for TOPAZ
CreateSimulationWorkspace(
    Instrument='TOPAZ',
    BinParams='1,100,10000',
    UnitX='TOF',
    OutputWorkspace='cws',
)
cws = mtd['cws']

# Set the UB matrix for the sample
# u, v is the critical part, we can start with the
# ideal position
SetUB(
    Workspace="cws",
    u='1,0,0',  # vector along k_i, when goniometer is at 0
    v='0,1,0',  # in-plane vector normal to k_i, when goniometer is at 0
    **lc_natrolite,
)

# Generate predicted peak workspace
dspacings = convert({'min': 1.0, 'max': 10.0})
wavelengths = convert({'min': 0.8, 'max': 2.9})

# Collect peaks over a range of omegas
CreatePeaksWorkspace(OutputWorkspace='pws')
omegas = range(0, 180, 3)

for omega in omegas:
    SetGoniometer(
        Workspace="cws",
        Axis0=f"{omega},0,1,0,1",
    )

    PredictPeaks(
        InputWorkspace='cws',
        WavelengthMin=wavelengths.min,
        wavelengthMax=wavelengths.max,
        MinDSpacing=dspacings.min,
        MaxDSpacing=dspacings.max,
        ReflectionCondition='All-face centred',
        OutputWorkspace='_pws',
    )

    CombinePeaksWorkspaces(
        LHSWorkspace="_pws",
        RHSWorkspace="pws",
        OutputWorkspace="pws",
    )

IndexPeaks("pws")

CloneWorkspace(InputWorkspace='pws', OutputWorkspace='clone')

MoveInstrumentComponent(
    Workspace='clone',
    ComponentName='moderator',
    X=0, Y=0, Z=-0.01,
    RelativePosition=True,
)


MoveInstrumentComponent(
    Workspace='clone',
    ComponentName='sample-position',
    X=0.002, Y=-0.003, Z=-0.003,
    RelativePosition=True,
)

SCDCalibratePanels(
        PeakWorkspace="clone",
        a=natrolite.a,
        b=natrolite.b,
        c=natrolite.c,
        alpha=natrolite.alpha,
        beta=natrolite.beta,
        gamma=natrolite.gamma,
        CalibrateL1=True,
        CalibrateBanks=False,
        CalibrateT0=True,
        ToleranceT0=0.01,
        SearchRadiusT0=t0 * 2,
        TuneSamplePosition=True,
        OutputWorkspace="testCaliTable",
        XmlFilename="/tmp/test.xml",
    )

def shift_by_t0(input_workspace, t0):
    r"""
    @brief Shift the TOF of all peaks by T0
    @details The contents of the following columns can be changed: 'Wavelength', 'QSample', 'QLab'.
    Other columns ('TOF', 'Energy', 'DSpacing') are read-only.
    @param t0: shift in TOF, in micro-seconds
    """
    w = mtd[str(input_workspace)]
    tofs = w.column('TOF')
    waves = w.column('Wavelength')
    qlabs = w.column('QLab')
    qsamples = w.column('QSample')
    for i in range(w.rowCount()):
        new_wave = waves[i] * (tofs[i] + t0) / tofs[i]
        w.setCell('Wavelength', i, new_wave)
        w.setCell('QLab', i, qlabs[i] * (waves[i] / new_wave))
        w.setCell('QSample', i, qsamples[i] * (waves[i] / new_wave))

for t0 in (1., 10., 100.):
    CloneWorkspace(InputWorkspace='pws', OutputWorkspace='clone')
    shift_by_t0('clone', t0)

    SCDCalibratePanels(
            PeakWorkspace="clone",
            a=natrolite.a,
            b=natrolite.b,
            c=natrolite.c,
            alpha=natrolite.alpha,
            beta=natrolite.beta,
            gamma=natrolite.gamma,
            CalibrateL1=False,#True,
            CalibrateBanks=False,
            CalibrateT0=True,
            ToleranceT0=0.01,
            SearchRadiusT0=t0 * 2,
            TuneSamplePosition=False, #True,
            OutputWorkspace="testCaliTable",
            XmlFilename="/tmp/test.xml",
        )
```

---

#### Reviewer ####

Please comment on the following ([full description](http://developer.mantidproject.org/ReviewingAPullRequest.html)):

##### Code Review #####

- Is the code of an acceptable quality?
- Does the code conform to the [coding standards](http://developer.mantidproject.org/Standards/)?
- Are the unit tests small and test the class in isolation?
- If there is GUI work does it follow the [GUI standards](http://developer.mantidproject.org/Standards/GUIStandards.html)?
- If there are changes in the release notes then do they describe the changes appropriately?

##### Functional Tests #####

- Do changes function as described? Add comments below that describe the tests performed?
- Do the changes handle unexpected situations, e.g. bad input?
- Has the relevant (user and developer) documentation been added/updated?

Does everything look good? Mark the review as **Approve**. A member of `@mantidproject/gatekeepers` will take care of it.

### CODE REVIEW

**Comment by jmborr** [comment]
@KedoKudo For this part of your workbench test:
```python
SCDCalibratePanels(
        PeakWorkspace="pws",
        a=silicon.a,
        b=silicon.b,
        c=silicon.c,
        alpha=silicon.alpha,
        beta=silicon.beta,
        gamma=silicon.gamma,
        CalibrateL1=True,
        CalibrateBanks=False,
        CalibrateT0=True,
        TuneSamplePosition=True,
        OutputWorkspace="testCaliTable",
        XmlFilename="test.xml",
    )
```
can you define namedtuple `silicon`?

**Comment by KedoKudo** [comment]
> @KedoKudo For this part of your workbench test:
> 
> ```python
> SCDCalibratePanels(
>         PeakWorkspace="pws",
>         a=silicon.a,
>         b=silicon.b,
>         c=silicon.c,
>         alpha=silicon.alpha,
>         beta=silicon.beta,
>         gamma=silicon.gamma,
>         CalibrateL1=True,
>         CalibrateBanks=False,
>         CalibrateT0=True,
>         TuneSamplePosition=True,
>         OutputWorkspace="testCaliTable",
>         XmlFilename="test.xml",
>     )
> ```
> 
> can you define namedtuple `silicon`?

Sorry about the confusion, should have updated the function call to use natrolite.
The testing script in the description is not updated.

**Comment by jmborr** [comment]
I don't understand the output I'm getting. I moved both the moderator and the sample position like this:
```python
CloneWorkspace(InputWorkspace='pws', OutputWorkspace='clone')

MoveInstrumentComponent(
    Workspace='clone',
    ComponentName='moderator',
    X=0, Y=0, Z=-0.01,
    RelativePosition=True,
)


MoveInstrumentComponent(
    Workspace='clone',
    ComponentName='sample-position',
    X=0.002, Y=-0.003, Z=0,
    RelativePosition=True,
)


SCDCalibratePanels(
        PeakWorkspace="clone",
        a=natrolite.a,
        b=natrolite.b,
        c=natrolite.c,
        alpha=natrolite.alpha,
        beta=natrolite.beta,
        gamma=natrolite.gamma,
        CalibrateL1=True,
        CalibrateBanks=False,
        CalibrateT0=True,
        TuneSamplePosition=True,
        OutputWorkspace="testCaliTable",
        XmlFilename="/tmp/test.xml",
    )
```
And the log output seems to indicate that the sample position is calibrated _twice_

> ** Calibrating L1 (moderator) as requested
> -- Fit L1 results using 11076 peaks:
>     dL1: 0.010362 
>     L1 18.01 -> 17.9996 
>     dT0 = -0.154085 
>     dSamplePos = (-0.00199842,0.00300038,7.27082e-07)
>     chi2/DOF = 2.45087e-06
> ** Tunning sample position as requested
> -- Tune SamplePos = (0,0,0) is below tolerance (1e-06), zero it
> -- Tune SamplePos results using 11076 peaks:
>   deltaSamplePos = (0,0,0)
>   chi2/DOF = 2.45072e-06
> -- Generate calibration table
> Generate a TableWorkspace to store calibration results.
> finished generating tables
> Generating xml tree 
> 	Saving parameter file as /tmp/test.xml
> Saving DetCal file in SCDCalibrate2.DetCal

Is this supposed to happen?

**Comment by KedoKudo** [comment]
> I don't understand the output I'm getting. I moved both the moderator and the sample position like this:
> 
> ```python
> CloneWorkspace(InputWorkspace='pws', OutputWorkspace='clone')
> 
> MoveInstrumentComponent(
>     Workspace='clone',
>     ComponentName='moderator',
>     X=0, Y=0, Z=-0.01,
>     RelativePosition=True,
> )
> 
> 
> MoveInstrumentComponent(
>     Workspace='clone',
>     ComponentName='sample-position',
>     X=0.002, Y=-0.003, Z=0,
>     RelativePosition=True,
> )
> 
> 
> SCDCalibratePanels(
>         PeakWorkspace="clone",
>         a=natrolite.a,
>         b=natrolite.b,
>         c=natrolite.c,
>         alpha=natrolite.alpha,
>         beta=natrolite.beta,
>         gamma=natrolite.gamma,
>         CalibrateL1=True,
>         CalibrateBanks=False,
>         CalibrateT0=True,
>         TuneSamplePosition=True,
>         OutputWorkspace="testCaliTable",
>         XmlFilename="/tmp/test.xml",
>     )
> ```
> 
> And the log output seems to indicate that the sample position is calibrated _twice_
> 
> > ** Calibrating L1 (moderator) as requested
> > -- Fit L1 results using 11076 peaks:
> > dL1: 0.010362
> > L1 18.01 -> 17.9996
> > dT0 = -0.154085
> > dSamplePos = (-0.00199842,0.00300038,7.27082e-07)
> > chi2/DOF = 2.45087e-06
> > ** Tunning sample position as requested
> > -- Tune SamplePos = (0,0,0) is below tolerance (1e-06), zero it
> > -- Tune SamplePos results using 11076 peaks:
> > deltaSamplePos = (0,0,0)
> > chi2/DOF = 2.45072e-06
> > -- Generate calibration table
> > Generate a TableWorkspace to store calibration results.
> > finished generating tables
> > Generating xml tree
> > Saving parameter file as /tmp/test.xml
> > Saving DetCal file in SCDCalibrate2.DetCal
> 
> Is this supposed to happen?

TL;DR; Yes

Longer version:
L1, T0 and sample position are dependent on each other, therefore it is better to have the optimizer optimize them together instead of one at a time (something I noticed while working on this user story).
So the first log is the one doing the heavy lifting as it finds the answer, and the second one is more of a sanity check that the sample position found along with L1 is already good enough.

**Comment by jmborr** [comment]
Well, if this threw me into confusion then it's highly likely other users will be also confused with looking at the log messages.  You may want to rephrase a bit the log output so that what the log states is consistent with what the algorithm does.

also, what are the units of `dT0`? I see `dT0 = -0.154085` but is this micro-seconds?

**Comment by KedoKudo** [comment]
> Well, if this threw me into confusion then it's highly likely other users will be also confused with looking at the log messages. You may want to rephrase a bit the log output so that what the log states is consistent with what the algorithm does.
> 
> also, what are the units of `dT0`? I see `dT0 = -0.154085` but is this micro-seconds?

yes, dT0 is in micro seconds.

**Comment by jmborr** [comment]
> > Well, if this threw me into confusion then it's highly likely other users will be also confused with looking at the log messages. You may want to rephrase a bit the log output so that what the log states is consistent with what the algorithm does.
> > also, what are the units of `dT0`? I see `dT0 = -0.154085` but is this micro-seconds?
> 
> yes, dT0 is in micro seconds.

Could you write the units after the number, in the log message?

**Comment by jmborr** [comment]
If I change the values in the `T0` column of the `pws` `PeaksWorkspace`, will `SCDCalibratePanels` pick these values or will it just use with the wavelength and ignore the `T0` values? I'm trying to figure out a way to test for `T0`

**Comment by KedoKudo** [comment]
> > > Well, if this threw me into confusion then it's highly likely other users will be also confused with looking at the log messages. You may want to rephrase a bit the log output so that what the log states is consistent with what the algorithm does.
> > > also, what are the units of `dT0`? I see `dT0 = -0.154085` but is this micro-seconds?
> > 
> > 
> > yes, dT0 is in micro seconds.
> 
> Could you write the units after the number, in the log message?

I will add the unit in the log message.

**Comment by KedoKudo** [comment]
> If I change the values in the `T0` column of the `pws` `PeaksWorkspace`, will `SCDCalibratePanels` pick these values or will it just use with the wavelength and ignore the `T0` values? I'm trying to figure out a way to test for `T0`

T0 is not a property of peaks, so you will have to update the wavelength of all peaks to get what you want.

**Comment by jmborr** [comment]
> T0 is not a property of peaks, so you will have to update the wavelength of all peaks to get what you want.

I forgot `T0` is calculated, not stored.  
If I'm not mistaken (can you verify?), I could change the wavelength according to this simple formula:

**Comment by KedoKudo** [comment]
> > T0 is not a property of peaks, so you will have to update the wavelength of all peaks to get what you want.
> 
> I forgot `T0` is calculated, not stored.
> If I'm not mistaken (can you verify?), I could change the wavelength according to this simple formula:
> 

It might be easier to just copy what [`LoadIsawDetcal`](https://github.com/mantidproject/mantid/blob/20aa5d47c871facd2fde76ff414b19872dd66101/Framework/DataHandling/src/LoadIsawDetCal.cpp#L214) did: using `ChangeBinOffset` to set T0.

**Comment by KedoKudo** [comment]
> > It might be easier to just copy what [`LoadIsawDetcal`](https://github.com/mantidproject/mantid/blob/20aa5d47c871facd2fde76ff414b19872dd66101/Framework/DataHandling/src/LoadIsawDetCal.cpp#L214) did: using `ChangeBinOffset` to set T0.
> 
> This should work with an EventsWorkspace containing diffraction data but as far as I can tell, your workbench script invokes `CreatePeaksWorkspace` which instantiates some peak objects without using any events or TOF histograms. Thus, there are no events of histograms to shift. The output of `CreateSimulationWorkspace` (workspace `csw`) is only used to fetch the instrument geometry, if I'm correct.
> 
> I see there are no test to verify that `T0` spits the correct shift when the diffraction data is shifted. I tried including this function in your script workbench script and then calibrate _only_ `T0`:
> 
> ```python
> CloneWorkspace(InputWorkspace='pws', OutputWorkspace='clone')
> 
> def shift_by_t0(input_workspace, t0):
>     r"""
>     @brief Shift the TOF of all peaks by T0
>     @details The contents of the following column can be changed: 'Wavelength'.
>     Other columns ('TOF', 'Energy', 'DSpacing') are read-only.
>     @param t0: shift in TOF, in micro-seconds
>     """
>     w = mtd[str(input_workspace)]
>     tofs = w.column('TOF')
>     waves = w.column('Wavelength')
>     for i in range(w.rowCount()):
>         new_wave = waves[i] * (tofs[i] + t0) / tofs[i]
>         w.setCell('Wavelength', i, new_wave)
> 
> shift_by_t0('clone', 90)
> 
> SCDCalibratePanels(
>         PeakWorkspace="clone",
>         a=natrolite.a,
>         b=natrolite.b,
>         c=natrolite.c,
>         alpha=natrolite.alpha,
>         beta=natrolite.beta,
>         gamma=natrolite.gamma,
>         CalibrateL1=False,#True,
>         CalibrateBanks=False,
>         CalibrateT0=True,
>         TuneSamplePosition=Fals, #True,
>         OutputWorkspace="testCaliTable",
>         XmlFilename="/tmp/test.xml",
>     )
> ```
> 
> I find that no matter what the value of `t0` I used (I used 30, 60, 90) I always obtain `dT0 = -9.9996`. The exception is `t0=0`, for which I get `dT0 = -0.0495253`. All this means either my shifting function is not correcting shifting TOF (I changed the wavelength but maybe I also have to change Qlab and Qsample) or the calibration of `T0` is incorrect.

dT0 is calculated based on Qlab, therefore updating QLab is necessary.
Otherwise SCDCalibratePanels2 cannot see the effect of T0.

**Review by jmborr — CHANGES_REQUESTED** [review]
> dT0 is calculated based on Qlab, therefore updating QLab is necessary.
> Otherwise SCDCalibratePanels2 cannot see the effect of T0.  
I have rescaled `QLab` and `QSample`  but still get the same `dT0`, irrespective of the `T0` shift.
```python
def shift_by_t0(input_workspace, t0):
    r"""
    @brief Shift the TOF of all peaks by T0
    @details The contents of the following column can be changed: 'Wavelength'.
    Other columns ('TOF', 'Energy', 'DSpacing') are read-only.
    @param t0: shift in TOF, in micro-seconds
    """
    w = mtd[str(input_workspace)]
    tofs = w.column('TOF')
    waves = w.column('Wavelength')
    qlabs = w.column('QLab')
    qsamples = w.column('QSample')
    for i in range(w.rowCount()):
        new_wave = waves[i] * (tofs[i] + t0) / tofs[i]
        w.setCell('Wavelength', i, new_wave)
        w.setCell('QLab', i, qlabs[i] * (waves[i] / new_wave))
        w.setCell('QSample', i, qsamples[i] * (waves[i] / new_wave))
```

File: Framework/API/src/FileFinder.cpp [hunk 1]
```diff
@@ -560,15 +564,31 @@ std::vector<std::string> FileFinderImpl::findRuns(const std::string &hintstr, co
...
     } else {
-      std::string path = findRun(*h, exts, useExtsOnly);
+      std::string path;
+      // Special check for "PG3", to cope with situation like '48314,48316'.
+      if (boost::algorithm::istarts_with(hint, "PG3")) {
+        if (h == hints.begin()) {
+          instrSName = "PG3";
+          path = findRun(*h, exts, useExtsOnly);
+        } else {
+          path = findRun(instrSName + *h, exts, useExtsOnly);
```
[the remark this answers is not in the document]
Comment (KedoKudo) [reply Framework/API/src/FileFinder.cpp:587]: This is not part of the calibration task.
Somehow Github thinks it is :man_shrugging:

File: Framework/Crystal/src/SCDCalibratePanels2.cpp [hunk 2]
```diff
@@ -95,30 +95,93 @@ void SCDCalibratePanels2::init() {
   setPropertySettings("gamma", std::make_unique<EnabledWhenProperty>("RecalculateUB", IS_DEFAULT));
 
   // Calibration options group
-  declareProperty("CalibrateT0", false, "Calibrate the T0 (initial TOF)");
+  // NOTE:
+  //  The general workflow of calibration is
```
[the remark this answers is not in the document]
Comment (KedoKudo) [reply Framework/Crystal/src/SCDCalibratePanels2.cpp:99]: Actually I did add recommendations in the documentation page that user should consider calibrate L1, T0 and sample position together if possible.
Given that different instrument will have their instrument specific needs (and this calibration alg is not restricted for one instrument), I don't really think we can provide a definitive workflow that everybody can follow.

[the remark this answers is not in the document]
Comment (jmborr) [reply Framework/Crystal/src/SCDCalibratePanels2.cpp:99]: I thought this was the workflow that was implemented in the code, which is why I was advocating to be included in the documentation. My bad.

**Comment by zjmorgan** [comment]
@jmborr Here is a Python script to adjust the TZERO offset of a peaks workspace.

```python
# import mantid algorithms, numpy and matplotlib
from mantid.simpleapi import *
import matplotlib.pyplot as plt
import numpy as np
from mantid.kernel import V3D

pws = LoadNexus(Filename='/SNS/TOPAZ/shared/stories/calibration/topaz_test_data.nxs', OutputWorkspace='pws')
LoadIsawUB('pws', '/SNS/TOPAZ/IPTS-24887/shared/2021A/scolecite_100K/use_Eng/SC100K_Monoclinic_C.mat')

DetID = []
TOF = []

N = mtd['pws'].getNumberPeaks()
for i in range(N):
    TOF.append(mtd['pws'].row(i)['TOF'])
    DetID.append(mtd['pws'].row(i)['DetID'])

LoadEmptyInstrument(InstrumentName='TOPAZ', OutputWorkspace='TOPAZ')
CreatePeaksWorkspace(InstrumentWorkspace='topaz', OutputWorkspace='cws', NumberOfPeaks=0)

h = 6.62607004e-34
mn = 1.674927471e-27
 
c = h/mn

L1 = mtd['cws'].getInstrument().getComponentByName('moderator').getPos().norm()

TZERO = 1.0

for i in range(N):
    det = mtd['cws'].getInstrument().getDetector(DetID[i]).getPos()
    L2 = det.norm()
    lamda = c*(TOF[i]+TZERO)/(L1+L2)*1e+4
    Q = V3D(-det.X()/L2*2*np.pi/lamda,-det.Y()/L2*2*np.pi/lamda,(1-det.Z()/L2)*2*np.pi/lamda)
    mtd['cws'].addPeak(mtd['cws'].createPeak(Q))
    mtd['cws'].getPeak(i).setGoniometerMatrix(mtd['pws'].getPeak(i).getGoniometerMatrix())
    mtd['cws'].getPeak(i).setRunNumber(mtd['pws'].getPeak(i).getRunNumber())
    mtd['cws'].getPeak(i).setIntHKL(mtd['pws'].getPeak(i).getIntHKL())
    HKL = mtd['pws'].getPeak(i).getHKL()
    mtd['cws'].getPeak(i).setHKL(HKL[0],HKL[1],HKL[2])
```

**Comment by jmborr** [comment]
> @jmborr Here is a Python script to adjust the TZERO offset of a peaks workspace.  

@zjmorgan  Are these peaks from Natrolite or some other sample. Below's the call to `SCDCalibratePanels` I'm using but it assumes the sample's Natrolite 
```python
    SCDCalibratePanels(
            PeakWorkspace="cws",
            a=natrolite.a,
            b=natrolite.b,
            c=natrolite.c,
            alpha=natrolite.alpha,
            beta=natrolite.beta,
            gamma=natrolite.gamma,
            CalibrateL1=False,#True,
            CalibrateBanks=False,
            CalibrateT0=True,
            ToleranceT0=0.01,
            SearchRadiusT0=TZERO * 2,
            TuneSamplePosition=False, #True,
            OutputWorkspace="testCaliTable",
            XmlFilename="/tmp/test.xml",
        )
```
And this is what I get for the correction `dT0`:

| input T0  (micro-s) | dT0 (micro-s)  |
|---:|---:|
| 1  |  |
| 10  |   |
|  100 | -9.99606  |

**Comment by zjmorgan** [comment]
@jmborr My mistake. This data is for scolecite a slightly different crystal.
```python
def convert(dictionary):
    return namedtuple('GenericDict', dictionary.keys())(**dictionary)

import os
directory = os.path.dirname(os.path.realpath(__file__))

lc_scolecite = {
    "a": 6.51201,  # A
    "b": 18.95454,  # A
    "c": 9.57822,  # A
    "alpha": 90,  # deg
    "beta": 108.9211 ,  # deg
    "gamma": 90,  # deg
}

scolecite = convert(lc_scolecite)
```

**Comment by KedoKudo** [comment]
> @jmborr My mistake. This data is for scolecite a slightly different crystal.
> 
> ```python
> def convert(dictionary):
>     return namedtuple('GenericDict', dictionary.keys())(**dictionary)
> 
> import os
> directory = os.path.dirname(os.path.realpath(__file__))
> 
> lc_scolecite = {
>     "a": 6.51201,  # A
>     "b": 18.95454,  # A
>     "c": 9.57822,  # A
>     "alpha": 90,  # deg
>     "beta": 108.9211 ,  # deg
>     "gamma": 90,  # deg
> }
> 
> scolecite = convert(lc_scolecite)
> ```

I think the c here is the incorrect one, here are the one from the corresponding user story
```
lc_scolecite = {
    "a": 6.51201,  # A
    "b": 18.95454,  # A
    "c": 9.75822,  # A
    "alpha": 90,  # deg
    "beta": 108.9211 ,  # deg
    "gamma": 90,  # deg
}

scolecite = convert(lc_scolecite)
```

**Comment by zjmorgan** [comment]
> I think the c here is the incorrect one, here are the one from the corresponding user story [SCD316](https://code.ornl.gov/sns-hfir-scse/diffraction/single-crystal/single-crystal-diffraction/-/issues/316)

That is true, however, the synthetic data is still generated with the incorrect data with `c=9.57`.

**Comment by jmborr** [comment]
@KedoKudo @zjmorgan  The final test

```python
######################################
# Perturb the source, sample, and T0
######################################
CloneWorkspace(InputWorkspace='pws', OutputWorkspace='clone')
MoveInstrumentComponent(
    Workspace='clone',
    ComponentName='moderator',
    X=0, Y=0, Z=-0.01,
    RelativePosition=True,
)
MoveInstrumentComponent(
    Workspace='clone',
    ComponentName='sample-position',
    X=0.002, Y=-0.003, Z=-0.003,
    RelativePosition=True,
)
shift_by_t0('clone', 42.0)
SCDCalibratePanels(
        PeakWorkspace="clone",
        a=natrolite.a,
        b=natrolite.b,
        c=natrolite.c,
        alpha=natrolite.alpha,
        beta=natrolite.beta,
        gamma=natrolite.gamma,
        CalibrateL1=True,
        ToleranceL1=0.0001,
        SearchRadiusL1=0.1,
        CalibrateBanks=False,
        CalibrateT0=True,
        ToleranceT0=0.01,
        SearchRadiusT0=100.0,
        TuneSamplePosition=True,
        ToleranceSamplePos=0.0001,
        SearchRadiusSamplePos=0.01,
        OutputWorkspace="testCaliTable",
        XmlFilename="/tmp/test.xml",
    )
```

results in the following adjustments:

```shell
** Calibrating L1 (moderator) as requested
-- Fit L1 results using 11076 peaks:
    dL1: 0.010362 
    L1 18.01 -> 17.9996 
    dT0 = -42.1541 
    dSamplePos = (-0.00199842,0.00300038,0.00300073)
    chi2/DOF = 2.45087e-06
** Tunning sample position as requested
-- Tune SamplePos = (0,0,0) is below tolerance (0.0001), zero it
-- Tune SamplePos results using 11076 peaks:
  deltaSamplePos = (0,0,0)
  chi2/DOF = 2.45072e-06
```

I believe they are pretty good. `L1` differs by 0.3mm and `T0` by 0.1 micro-second.
This PR is ready to be approved, unless any of you have objections to make.

**Comment by jmborr** [comment]
> That is true, however, the synthetic data is still generated with the incorrect data with `c=9.57`.

If I use this value for `c`, then the calibration becomes worse!

| input T0  (micro-s) | dT0 (micro-s)  |
|---:|---:|
| 1  | -1.99975 |
| 10  | -19.9971  |
|  100 | -127.774  |

I'm going to stick with the workbench script at the top, because it creates everything from scratch. This script shows that the calibration algorithm is working

**Review by jmborr — APPROVED** [review]
